### PR TITLE
library recommendations in curator

### PR DESCRIPTION
**Pull request by joshm1**
@stkem @yingjieMiao 

This is very much a WIP (parts of the code are obviously incomplete and features/tests are missing). Sharing anyway if you don't mind taking a quick look since the PR is already quite big.

### CODE REVIEW

**Comment by stkem** [comment]
Nice! Will review thoroughly today.

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/LibraryRecommendationGenerationCommander.scala [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      map(users => users.map(_.id.get).toSeq)
+
+  def getTopRecommendations(userId: Id[User], howManyMax: Int): Future[Seq[LibraryRecommendation]] = {
+    db.readOnlyReplicaAsync { implicit session =>
+      libraryRecRepo.getByTopMasterScore(userId, howManyMax)
+    }
+  }
+
+  private def shouldInclude(scoredLibrary: ScoredLibraryInfo): Boolean = {
+    scoredLibrary.masterScore > 0 // TODO(josh)
```
Comment (joshm1) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/LibraryRecommendationGenerationCommander.scala:43]: going to keep this at 0 for initial internal testing to get an idea of what kind of scores the current algorithm derives with real data

**Comment by joshm1** [comment]
@stkem @yingjieMiao  I think my first draft is done. I'm moving to the admin page next, but this can be deployed after review(s)

**Comment by yingjieMiao** [comment]
sorry, I was late on this. Will go through this tonight.

File: kifi-backend/modules/curator/app/com/keepit/curator/LibraryScoringHelper.scala [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+  }
+
+  private def getPopularityScore(candidate: CuratorLibraryInfo): Float = {
+    1 / (1 + Math.exp(-candidate.memberCount / 5)).toFloat
+  }
+
+  private def getSizeScore(candidate: CuratorLibraryInfo): Float = {
+    1 / (1 + Math.exp(-candidate.keepCount / 15)).toFloat
+  }
```
Comment (yingjieMiao) [inline kifi-backend/modules/curator/app/com/keepit/curator/LibraryScoringHelper.scala:118]: cortex also has 'entropy' info for a library. Big Library tends to have 'maxed entropy', in a 256 dim model, this is log_2(256) = 8.0

We could have some penalty for libraries with big entropy. (I can include this in the user/library score in the future)

**Comment by yingjieMiao** [comment]
lgtm overall. (didn't scrutinize though). It's probably better to push smaller pull requests. (i.e. push out self-contained modules early). 

:+1:

File: kifi-backend/modules/common/conf/evolutions/shoebox/267.sql [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# CURATOR
+
+# --- !Ups
+
+CREATE TABLE library_recommendation (
```
Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/267.sql:5]: Does not need to be in this PR, but this table should probably gain a few more fields similar to uri_recommendation, i.e. the topics of the library and possibly feedback fields ('clicked', 'followed', 'trashed' and such, these will depend on the final UI)

**Comment by stkem** [comment]
Nice work! Let's get it into prod :-)

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/LibraryRecommendationGenerationCommander.scala (hunk 1 above)
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/LibraryRecommendationGenerationCommander.scala:43]: :+1:

**Pull request merged**